src/MagnumPlugins/BasisImageConverter/BasisImageConverter.cpp

```diff
@@ -0,0 +1,216 @@
+    PARAM_CONFIG(multithreading, bool);
+    PARAM_CONFIG(seperate_rg_to_color_alpha, bool);
+    PARAM_CONFIG(disable_hierarchical_endpoint_codebooks, bool);
+
+    PARAM_CONFIG(hybrid_sel_cb_quality_thresh, float);
+
+    PARAM_CONFIG(global_pal_bits, int);
+    PARAM_CONFIG(global_mod_bits, int);
+
+    PARAM_CONFIG(endpoint_rdo_thresh, float);
+    PARAM_CONFIG(selector_rdo_thresh, float);
+
+    PARAM_CONFIG(mip_gen, bool);
+    PARAM_CONFIG(mip_renormalize, bool);
+    PARAM_CONFIG(mip_wrapping, bool);
+    PARAM_CONFIG(mip_srgb, bool);
+    PARAM_CONFIG(mip_scale, float);
+    PARAM_CONFIG(mip_smallest_dimension, int);
+    PARAM_CONFIG(mip_filter, std::string);
+
+    PARAM_CONFIG(max_endpoint_clusters, int);
+    PARAM_CONFIG(max_selector_clusters, int);
+    PARAM_CONFIG(quality_level, int);
+#undef PARAM_CONFIG
+
+    /* Without this, basis tries to find some global selector cookbook
+       and crashes on an assert that it does not exit */
+    params.m_no_hybrid_sel_cb = true;
+
+    /* Enable verbose debug output in basis to get reasons for failures */
+    basisu::enable_debug_printf(configuration().value<bool>("enable_debug_printf"));
+
+    if(image.size().x() <= 0 || image.size().y() <= 0) {
+        Error() << "Trade::BasisImageConverter::exportToData(): source image is empty";
+        return {};
+    }
+
+    /* Copy image data into the basis image. There is no way to construct a
+       basis image from existing data as it is based on a std::vector */
+    params.m_source_images.emplace_back(image.size().x(), image.size().y());
+
+    /* Get data properties and calculate the initial slice based on subimage
+       offset */
+    const std::pair<Math::Vector2<std::size_t>, Math::Vector2<std::size_t>> dataProperties =
+        image.dataProperties();
+    Containers::ArrayView<const char> inputData = image.data().suffix(
+        dataProperties.first.sum());
+
+    basisu::image& basisImage = params.m_source_images.back();
+
+    /* Do Y-flip and tight packing of image data */
+    const std::size_t rowSize = image.size().x()*image.pixelSize();
+    const std::size_t rowStride = dataProperties.second.x();
+    const std::size_t packedDataSize = rowSize*image.size().y();
+
+    /* basis image is always RGBA, fill in alpha if necessary */
+    if (image.format() == PixelFormat::RGB8Unorm) {
+        const std::size_t width = image.size().x();
+        auto basisView = Containers::arrayCast<Color4ub>(
+            Containers::arrayView<basisu::color_rgba>(basisImage.get_ptr(), width));
+
+        for(Int y = 0; y < width; ++y) {
+            const auto src = Containers::arrayCast<const Color3ub>(inputData
+                .suffix(y*rowStride).prefix(rowSize));
+
+            Color4ub* destPixel = basisView.suffix(width).begin();
+            for(const Color3ub& pixel: src) {
+                destPixel->rgb() = pixel;
+                destPixel->a() = 255;
+            }
+        }
+
+    } else {
```

Comment:
> as a sanity check in case the format support gets extended for HDR formats and such.

Uhm, but I have a check for the format at the very top? Is that insufficient?

Reply:
Re the 1/2-component images: checking basis code etc, I think this can be achieved by converting the input to RRR(G) and compressing that.

Reply:
> Uhm, but I have a check for the format at the very top? Is that insufficient?

Those two checks are (relatively) far from each other, so if you extend the one at the top but forget this one at the bottom, it'll silently produce wrong results (instead of blowing up). That's what I meant.

> I think this can be achieved by converting the input to RRR(G) and compressing that.

Hmm, that sounds like I should extend the importer tests with those as well.

Reply:
Ah! Makes a lot of sense now, thanks for explaining, will add the assert 👍